### DISCUSSION TITLE
Add a new cropping mode to keep the reference to the original image

### DISCUSSION THREAD

**Discussion by qborisb** [directus/directus#14578] [Feature Requests]
We would really appreciate one extra feature to the image editor. Let me show you how it would work by the following example :smile:

In the image editing modal, we want to add a new option called `Create New Image`:

![Screenshot 2022-07-22 at 14 33 54](https://user-images.githubusercontent.com/54835674/180440060-6557e25f-e213-4dfb-9266-2ac5948d52a6.png)

If you check it and save the changes, a new image (crop) will be created on the disk while keeping a reference to the original image and the crop coordinates. If the crop was created within the settings of an item, then the item would now need to have a relationship with the crop, not the original image.

![Screenshot 2022-07-22 at 14 36 18](https://user-images.githubusercontent.com/54835674/180440502-861f1fdd-3fff-4837-aa79-101832929803.png)

Then, when you want to edit the image again, the image editor will recognize that it is a crop referencing a certain original image. This original image will be used for cropping, while the crop will be used as a frame that will get automatically drawn when the modal gets open:

![Screenshot 2022-07-22 at 14 40 54](https://user-images.githubusercontent.com/54835674/180441209-8c9f0400-2201-4662-880d-d12eb6a40b15.png)

Importantly, when the `Create New Image` checkbox remains unchecked, the behaviour of the image editor is then unchanged from what is currently implemented in Directus.

The following video will better illustrate the feature we have in our mind:

https://user-images.githubusercontent.com/54835674/180442483-0cc80a00-7a89-4dde-8843-98c7a4babc25.mp4

**Comment by rijkvanzanten** [comment]
What is the "top level" use case for this? Is it to select the vocal focus point in an image for use?

**Reply by dvic** [reply]
> What is the "top level" use case for this? Is it to select the vocal focus point in an image for use?

In addition to what @qborisb mentioned below: usually we have clients that have some sort of "master" photo of an item from which various crops are made for different ui elements (slides in a slideshow, banner images, etc.). So the usecase here is also to have to original present in the media library without having to upload it again (because each crop is a separate image).

**Reply by dvic** [reply]
this opens up new possibilities for directus (e.g., to have the media library to be more of a digital asset management kind of thing)

**Comment by qborisb** [comment]
@rijkvanzanten no, the use case would be to simplify re-editing images. If crops keep the reference to the original image, it makes it much easier to adjust them again relative to the original whenever users want to make slight changes to them.

**Reply by rijkvanzanten** [reply]
Gotcha, so you're effectively looking to "undo" the previously made edit. The current workflow effectively replaces the image file contents with the updated version. I'm wondering if instead we should save the made edits to the database, and apply them on the fly when the asset is requested through `/assets`. That way, you can make any edit (and undo any change to that edit) while still maintaining only a single copy of the file 🤔

**Reply by qborisb** [reply]
This is fine by us! Seems like what we want is indeed supported by sharp.

**Reply by dvic** [reply]
@rijkvanzanten To summarise and check we're on the same page: each time a new crop is made, you want to duplicate the row in  "directus_files" but add crop coordinates to it. Then the logic of re-using and "overwriting" the crops can remain the same as described above.

However, there is one small thing here: if we obtain a crop of a particular `directus_files` row, then that row itself must be locked for inline cropping, because otherwise all the other crops will become invalid because they are using its storage. Do you agree with this?

**Reply by rijkvanzanten** [reply]
No, I wouldn't want to create duplicate rows in directus_files, as that also gets into syncing problems with the other fields. 

I was thinking we could maybe add a `default transformation` column to directus_files, that effectively sets the transformations used in the `/assets` endpoint when no other transformations are passed. Alongside that, we can add a `?original` query param to `/assets` to fetch the original file without any transformations applied. That way, rather than saving a whole new image (or duplicate directus_files), the "image edit" functionality effectively just sets what default transformation to apply when fetching the asset, which in turn means that every image transformation is non-destructive by default 🤔

**Reply by dvic** [reply]
And how would this work when you have multiple crop variants (say, thumbnail, heading, banner) and you want to re-use these across collections? Put all these variants into a single `directus_files` row as a jsonb column?

**Reply by rijkvanzanten** [reply]
That's already possible using the regular presets functionality 👍🏻 This would only control the "default preset" used when accessing the file without any transformations in the browser. So for example, currently `/assets/:uuid` would return the original asset. My idea above would effectively allow you to control what `/assets/:uuid` returns by default, with an additional `?original` to get back to the original file data.

**Reply by dvic** [reply]
Yes, but we still need a way for a user  to "pick" a crop for a particular collection field, right? I want users of directus to be able to pick this "preset" and not the application serving the data in directus.

**Reply by rijkvanzanten** [reply]
Ohh, you're trying to alter the image per image field and store the alterations linked to the original item? I think I'm not fully grasping the use case / final needs here 😁

**Reply by qborisb** [reply]
Here is what we are trying to achieve (from user's perspetive):

As a user, I want to be able to create crops from any uploaded image / image chosen from the files library. Then, once the crop is created, I expect that this crop can be reused in other items of any collection (i.e., it can be selected from the files library on other fields). Importantly:
1. The original image should still remain accessible - it should not be overwritten by the newly created crop (currently the underlying file row is being overwritten).
2. The crop should keep the reference to the original image, so that it can be easily adjusted with respect to the original.
3. "Cropping" the original should not affect any other crops that are based of it

We already did some work to check whether it is possible to achieve something like that in Directus (see [this draft PR](https://github.com/qborisb/directus/pull/1)). This PR works as follows:
- It adds 2 new fields to the files table: `crop_original_image_id`and `crop_coordinates`. 
- It adds a new checkbox to the image editor called "Create New Image" (as shown in the image all the way on top of this discussion). **When the user checks it** and saves the changes, a new row in files table gets created  - a crop with 2 aforementioned fields set to `id` of the original and coordinates of the crop. If the user **leaves the checkbox untouched**, the behaviour of the editor remains as it was before - it overwrites the underlying file, while leaving `crop_original_image_id`and `crop_coordinates` empty.

By creating a new row in the `files` table, we are able to partially satisfy the user story: newly created crops can be reused within other image fields. However, current PR does not satisfy the third requirement. We tried to think how to solve this, but in the end the best solution is to not physically crop the files at all, but use the assets endpoint for this purpose (i.e., all crops in the system will become "soft crops"). But we still need to have separate rows for `directus_files` so that we can reference these crops from various collections.

**Reply by rijkvanzanten** [reply]
I can't say I remember ever seeing a similar request before 🤔 Do you have a user story that aligns with the output usage to help me understand where these crops are then used? So far, the workflow I'm used to seeing is that various asset presets are setup globally in settings, and then selected in the output usage (through the `key` query param on `/assets`), rather than setting that on a per-image level. The "select a focal point in the image" case has come up before, but it definitely sounds like there's a specific project where this is required that I'd love to learn more about 😁

**Reply by dvic** [reply]
There is indeed a user story behind this :)

The approach described above is something that we often use for projects where we build a website for a client and provide a custom CMS that goes along with it. This is the outline of it:

- We have many clients in the theatre industry and it's very common for them to get "official" high-resolution images from the producers of the shows they are selling. These images are various shots of the artist and / or stills of the show while being performed by the artist.
- The websites that we build consist of various UI components that can be configured in the CMS, by among other things, setting an image (think of slide shows, item listings, landing page headers, etc.).
- However, most of these components require certain aspect ratios of the images used and/or it's desirable to have a nice focal point. For example, a particular listing item component may require a square thumbnail of a show, while the heading component of the detail page of that show requires an aspect ratio of 16:9.
- In our application that serves the website we do use a service similar to `sharp` to on-the-fly *scale* these images to generate [image src-sets](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images), but we want the users in the CMS to decide on the focal point given the aspect ratio requirement.
- Finally, the users might change their minds of a particular chosen focal point and they want to easily edit this (so the original high-res image should not be overwritten).

I hope this clears it up a bit :)

**Reply by dvic** [reply]
And by the way, this feature is one of the most loved features of our platform, it saves our clients *a lot* of time: they don't have the re-upload images each time they want a different focal point.

**Reply by qborisb** [reply]
@rijkvanzanten so, what do you think about implementing this feature? Can we schedule a short meeting to discuss it? I can show a demo of how it would look like!

**Reply by rijkvanzanten** [reply]
I'd love to discuss it more! From the points above, it definitely sounds like the initial design goal would be a (re)configurable vocal point which is stored on the file metadata, and can be edited after the fact. The API would then automatically use this vocal point as the center point in its various cropping operations (which can then be overridden in another query param). I do think that the actual cropping sizes etc are better of controlled through the API still, but happy to keep discussing that too. @benhaynes thoughts on everything above?

**Reply by benhaynes** [reply]
I **really** like the idea of adding focal point coordinate metadata to the file library. It seems pretty straightforward to get this added, but I wouldn't say trivial. One thing to figure out would be how to regenerate existing thumbnails if the focal point changes later on. Would love to work on a design for this when we're ready.

I'd be happy to hop on a call to discuss, I'm also curious if this is something you or your company is looking to sponsor or submit a pull request for. Just to set expectations, our engineering team is pretty maxed out at the moment, and so even if approved our core team wouldn't get to this for a bit.

Also, we might want to pull this into a separate GitHub discussion, I thought there was a focal point one somewhere. This seems related but different to the aspect ratio topic above.

**Reply by janwillemvd** [reply]
Hi @benhaynes, I'm a co-worker of @qborisb and @dvic. 

> I'd be happy to hop on a call to discuss, I'm also curious if this is something you or your company is looking to sponsor or submit a pull request for. 

Yes, definitely! We'd love to submit a PR request for this :) In fact, we have a proof of concept ready to demonstrate the idea. We agree with you it's relatively straightforward to add this, but not trivial and we should make sure we're on the same page ;) 

Maybe we can already schedule a call to discuss it / show the proof of concept?

**Reply by benhaynes** [reply]
Absolutely. Are you on our Discord? If you DM there I will give you my calendly link and we can coordinate a chat (with Rijk).

https://directus.chat

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
